tools/audio/extraction/audio_extract.py

```diff
@@ -0,0 +1,267 @@
+#!/usr/bin/env python3
+# audio_extract.py
+# SPDX-FileCopyrightText: © 2024 ZeldaRET
+# SPDX-License-Identifier: CC0-1.0
+#
+#   Extract audio files
+#
+
+import argparse, os
+from multiprocessing.pool import ThreadPool
+from typing import Dict, List, Tuple, Union
+from xml.etree import ElementTree
+from xml.etree.ElementTree import Element
+
+from audio_tables import AudioCodeTable, AudioStorageMedium
+from audiotable import AudioTableFile
+from audiobank_file import AudiobankFile
+
+from util import debugm, error, incbin
+
+from config import GAMEVERSION_ALL_OOT, VERSION_TABLE, GameVersionInfo
+from config import AUDIOTABLE_BUFFER_BUGS, FAKE_BANKS
+
+BASEROM_DEBUG = False
+
+# ======================================================================================================================
+#   Run
+# ======================================================================================================================
+
+def collect_sample_banks(rom_image : memoryview, extracted_dir : str, version_info : GameVersionInfo,
+                         table : AudioCodeTable, samplebank_xmls : Dict[int, Tuple[str, Element]]):
+    sample_banks = []
+
+    for i,entry in enumerate(table):
+        entry : AudioCodeTable.AudioCodeTableEntry
+
+        assert entry.short_data1 == 0 and entry.short_data2 == 0 and entry.short_data3 == 0, \
+               "Bad data for Sample Bank entry, all short data should be 0"
+        assert entry.medium == AudioStorageMedium.MEDIUM_CART , \
+               "Bad data for Sample Bank entry, medium should be CART"
+
+        if entry.size == 0:
+            # Pointer to other entry, in this case the rom address is a table index
+
+            entry_dst = table.entries[entry.rom_addr]
+            sample_banks[entry.rom_addr].register_ptr(i)
+            sample_banks.append(entry_dst.rom_addr)
+
+            # debugm(f"{i} Pointer: {entry.rom_addr} -> 0x{entry_dst.rom_addr:X}, 0x{entry_dst.size:X}")
+        else:
+            # Check whether this samplebank suffers from the buffer bug
+            # TODO it should be possible to detect this automatically by checking padding following sample discovery
+            if version_info.version_id in AUDIOTABLE_BUFFER_BUGS:
+                bug = i in AUDIOTABLE_BUFFER_BUGS[version_info.version_id]
+            else:
+                bug = False
+
+            bank = AudioTableFile(i, rom_image, entry, version_info.audiotable_rom + table.rom_addr, buffer_bug=bug,
+                                  extraction_xml=samplebank_xmls.get(i, None))
+
+            if BASEROM_DEBUG:
+                bank.dump_bin(f"{extracted_dir}/baserom_audiotest/audiotable_files/{bank.file_name}.bin")
+
+            sample_banks.append(bank)
+
+    return sample_banks
+
+def bank_data_lookup(sample_banks : List[AudioTableFile], e : Union[AudioTableFile, int]) -> AudioTableFile:
```

Comment:
The recursion/Union here is a bit weird. Is this ever called with an AudioTableFile anyway? The only two uses I see are in bank_data_lookup and they're both passed ids instead of an AudioTableFile object

Reply:
My type annotation was wrong for `sample_banks`, it's also a `List[Union[AudioTableFile, int]]` since it can either contain a "real" bank or a "pointer" to a bank (the rom address is instead an index identifying another table entry)